Abstract node side functionality of attachable plugins

swagiaal opened the pull request:
- Create PhysicalAttacher interface to abstract MountDevice and
  WaitForAttach.
- Create PhysicalDetacher interface to abstract WaitForDetach and
  UnmountDevice.
- Expand unit tests to check that Attach, Detach, WaitForAttach,
  WaitForDetach, MountDevice, and UnmountDevice get call where
  appropriet.

Physical{Attacher,Detacher} are working titles suggestions welcome. Some other thoughts:
- NodeSideAttacher or NodeAttacher.
- AttachWatcher
- Call this Attacher and call the Current Attacher CloudAttacher.
- DeviceMounter (although there are way too many things called Mounter right now :/)

This is to address: https://github.com/kubernetes/kubernetes/pull/21709#issuecomment-192035382

@saad-ali

saad-ali commented on pkg/volume/volume.go:
```diff
@@ -132,11 +132,39 @@ type Attacher interface {
 }
 
+type PhysicalAttacher interface {
+	// WaitForAttach examines the file system for the device; if
```
Don't suggest implementation in the GoDoc for an interface method. Meaning don't say *how* this method should be implemented (e.g. "examines the file system"), instead focus on *what* this method should do (e.g. "WaitForAttach will block until the volume has been successfully attached to this machine. If the volume is not attached within X minutes, it will return an error").

saad-ali commented on pkg/volume/volume.go:
```diff
@@ -132,11 +132,39 @@ type Attacher interface {
+	// returns otherwise it polls until the device appears. If the
+	// device is not attached within the timeout period an error
+	// is returned.
+	WaitForAttach() error
```
Maybe pass the timeout period as a parameter so that it can be controlled in a common place rather than as a one-off for each plugin. Honestly, I could go either way on this: it could be argued that each plugin should define what this period is independently (since different types of volumes could take a different amount of for this operation to complete).

saad-ali replied:
Also, consider having this method return the path for the newly attached device (i.e. `makeGlobalPDName(...)` for GCE so that the calling code can use it for the bind mount step.

swagiaal replied:
> Also, consider having this method return the path for the newly attached device (i.e. makeGlobalPDName(...) for GCE so that the calling code can use it for the bind mount step.

The device path is different from the global mount point. We would need both. So I'll have `WaitForAttach` return the device path and `GetDeviceMountPath()` as suggested below.

If I make those changes I don't think we'll need `MountDevice()` anymore the kubelet can just mount the device to the global path using its own mounter.

saad-ali replied:
That sounds good to me.

swagiaal replied:
> If I make those changes I don't think we'll need MountDevice() anymore the kubelet can just mount the device to the global path using its own mounter.

Okay changing my mind about this; the global mount has minor variations between volume plugins. Some inspect the volume spec to check for read only volumes, and RBD implements fencing etc.

saad-ali commented on pkg/volume/volume.go:
```diff
@@ -132,11 +132,39 @@ type Attacher interface {
 	Attach() error
 }
 
```
Comment on `GetPath()` above:
```GO
type Volume interface {
	// GetPath returns the directory path the volume is mounted to.
	GetPath() string

	// MetricsProvider embeds methods for exposing metrics (e.g. used,available space).
	MetricsProvider
}
```

We need to make a clear distinction between the global mount path and the bind mount path. So let's break `GetPath()` in to `GetDeviceMountPath()` `GetPodBindMountPath()`

saad-ali replied:
Comments on both should make it clear that the paths may not necessarily exist, just what the expected location is.

swagiaal replied:
hmm.. The global vs bind concept is only applicable to attachable volumes. So how about I add `GetDeviceMountPath()` to `PhysicalAttacher` and leave `GetPath()` as is but expand the comment ?

saad-ali replied:
Good point. Actually both only make sense on kubelet. So it would make sense to move both to `PhysicalAttacher`

swagiaal replied:
Hmm.. only one can be moved I think because GetPath() is applicable to all volumes not just attachable ones

saad-ali commented:
> Physical{Attacher,Detacher} are working titles suggestions welcome.

Do you plan to remove the existing Mounter/Unmounter interfaces in a separate PR? I ask because, honestly I like those names the best. `Attacher`/`Detacher`/`Mounter`/`Unmounter`--clean and to the point. My second choice would be `DeviceMounter`. But then we should rename the current `Attacher` to `DeviceAttacher` for consistency. Prefixing `Device` seems unnecessary considering they are already part of the `volume` package so it is clear what is being attached, mounted, etc.

swagiaal commented:
> Do you plan to remove the existing Mounter/Unmounter interfaces in a separate PR? I ask because, honestly I like those names the best. Attacher/Detacher/Mounter/Unmounter

I don't plan to do any renaming. These interfaces will be in addition to the ones you list above. We need Attacher to be execution location independent and talk to the cloud provider and PhysicalAttacher to be node dependent and examine the node to confirm the attachment.

Come to think of it... if we remove `MountDevice` from `PhysicalMounter` that opens up some other name options like `AttachWatcher` or `AttachObserver`

saad-ali commented:
The existing `Mounter`/`Unmounter` will be deprecated right? What are the final set of interfaces that you see once this refactor is all done?

swagiaal commented:
> The existing Mounter/Unmounter will be deprecated right?

Hmm.. I was not planning on it.

> What are the final set of interfaces that you see once this refactor is all done?

Here is what I had in mind: 

```go
// Mounter mounts mounts a volume for a pod
// implemented by attachable an non attachable storage.
type Mounter interface {
}

type Unmounter interface {
}

// Attacher talks to the cloud provider to request attaching a volume
// to a node. Implemented by attachable volumes only. Must be runnable
// from either node or master.
type Attacher interface {
}

type Detacher interface {
}

// PhysicalAttacher -or some other better name- watches for the device
// to appear on the node after an attach call has been made to the cloud provider.
// implemented by attachable volumes only and must run on node.
type PhysicalAttacher interface {
}

type PhysicalDetacher interface {
}
```

saad-ali commented:
The existing `Mounter` currently requires `SetUp()` and `SetUpAt()` methods. These methods are responsible for a hodgepodge of things including attaching, mounting device, and bind mounting pod. The new `PhysicalAttacher` (or whatever we will call it) defines `MountDevice()` and the wait for attach stuff. I guess my confision is what is the difference between `SetUp/SetUpAt` and `MountDevice`? My understanding was that we were going to eliminate `SetUp/SetUpAt` as part of this refactor because it was a bad/leaky abstraction and replace it with clearer concepts.

swagiaal commented:
So the way I see it is
`MountDevice` mounts the device to a global path
`SetUp/SetUpAt` bind mounts the global path to the pod path for attachable volumes or performs another kind of mount for non attachable volumes.
 
> what is the difference between SetUp/SetUpAt and MountDevice?

The difference is highlighted by non attachable volumes like nfs or glusterfs for example. In these volumes SetUp and SetUpAt are very custom and cannot be standardized. MountDevice applies to attachable volumes only which have the two level mount+bind mount setup.

> My understanding was that we were going to eliminate SetUp/SetUpAt as part of this refactor because it was a bad/leaky abstraction and replace it with clearer concepts.

We will make it thinner but I don't think we can eliminate it entirely because it looks very different among certain volume types. For attachable volumes we might be able to create one `Mounter` which can be reused by all attachable volumes.

The other option would be to have attachable volumes not implement Mounter/Unmounter and just do the operations in the Kubelet, but I think that is too ambitious I say lets leave them and if they all the implementations (for attachable volumes) turn out to be one liner call throughs to mounter/unmount then eliminate them.

swagiaal commented:
@saad-ali okay take a look at the latest commit (https://github.com/kubernetes/kubernetes/pull/24557/commits/019544daacf25972bdabf6248c76978e868c723d) I think it addresses all feedback and expresses my current understanding

saad-ali commented:
> The difference is highlighted by non attachable volumes like nfs or glusterfs for example. In these volumes SetUp and SetUpAt are very custom and cannot be standardized. MountDevice applies to attachable volumes only which have the two level mount+bind mount setup.

Got it. I propose something more radical: Plugins should not have to worry about mounting to pod specific location at all. Instead all plugins, weather they do attach or not, should only have to implement a `MountDevice()` method defining how to mount the volume to the global device mount path. Once this common device path exists the common volume code will be responsible for bind mounting this global location to the pod specific locations. I understand that volumes like NFS and GlusterFS don't follow this approach at the moment (instead they directly mount to the pod specific location and don't do any bind mounting), but is there any reason they couldn't follow this more generic "mount to generic location, and then let common code bind mount to pod specific location" approach?

The reason I advocate for this approach is because I want to simplify the design not continue to add more interfaces--I think reducing the set of things plugins need to implement will greatly simplify/cleanup the logic.

Thanks for bearing with me. I want to get this right so that the rest of the design falls cleanly out of this.

swagiaal commented:
Oh I see what you are saying. Technically speaking I don't think there is a reason why we can not do it but I don't think we should for a couple of reasons:
- It might make the interfaces more uniform but it introduces complexity to non attachable volumes. They now have an extra bind-mount step which they did not need. It would be like we are giving them work to do so that they can resemble our desired interfaces.
- It changes the way things are laid out on disk which, as you pointed out previously is a sort of API, which we have to maintain.
- lastly I am not sure it reduces the number of interfaces. You still need something to attach, something to wait for attach, and something to mount. Unless we combine the last two and make WaitForAttach a no-op where applicable.

Generally I find that refactoring works best when it is motivated by functionality rather than aesthetics.

> Thanks for bearing with me. I want to get this right so that the rest of the design falls cleanly out of this.

No worries at all!.. for what it is worth I think that we should use this PR to layout a general guide line and refine the interfaces as we incorporate more plugins into them. My feeling is that anything we come up with here will not survive contact with plugins unchanged :)

saad-ali commented:
> It might make the interfaces more uniform but it introduces complexity to non attachable volumes. They now have an extra bind-mount step which they did not need. It would be like we are giving them work to do so that they can resemble our desired interfaces.

To be clear, the bind-mount code would *not* be something that plugins would have to implement. It would be common volume code for all plugins. So it would actually make the plugins simpler, they would only need to specify how to mount the device to the global mount location, that's it--individual volume plugins wouldn't have to be aware of, implement, or do any bind mounting.

> It changes the way things are laid out on disk which, as you pointed out previously is a sort of API, which we have to maintain.

It would eliminate the `Setup` and `SetupAt` methods altogether. Plugins would only need to specify a MountDevice. AttachDevice and WaitForAttach would be optional if devices require attach.

> It changes the way things are laid out on disk which, as you pointed out previously is a sort of API, which we have to maintain.

Backwards compatibility is a good argument. If NFS and Gluster now have a global mount location, we'd need to write some code to make sure we are backwards compatible through upgrades which would increase the scope of this work. For that reason, I'm ok not going in that direction at this point. I still think it is something we should do but it doesn't have to be done now.

swagiaal commented:
So what is the final set of interfaces as you see them now ? I'll update the PR to match

saad-ali commented:
I'll take a look at what you have now in a couple hours and comment on it.

saad-ali commented:
Since `PhysicalAttacher` contains methods that are only for volumes that do attachment, might as well stick them inside the existing `Attacher` interface instead of creating another interface. What do you think?

And just to make sure I understand, will the existing `Mounter` methods `SetUp()`/`SetUpAt()` be called for all volumes or just non-attachable volumes? If all volumes, then will both `MountDevice` and `Setup` be called for attachable volumes? If so what will the separation of duties be? `MountDevice` does global mount. And `Setup` does pod bind mount?

swagiaal commented:
> Since PhysicalAttacher contains methods that are only for volumes that do attachment, might as well stick them inside the existing Attacher interface instead of creating another interface. What do you think?

hmm.. I kind of wanted to make the implementer of Attacher executable from both node and master. Do you think that is worth aiming for ?

> what will the separation of duties be? MountDevice does global mount. And Setup does pod bind mount?

That is what I was thinking yes. SetUp will also do a few other tasks such as the application of the fsGroup

saad-ali commented:
> hmm.. I kind of wanted to make the implementer of Attacher executable from both node and master. Do you think that is worth aiming for ?

I do. I think that is achievable by passing node/master specific params to the methods instead of initializing the object with them.

> That is what I was thinking yes. SetUp will also do a few other tasks such as the application of the fsGroup

Got it.

swagiaal commented:
> I do. I think that is achievable by passing node/master specific params to the methods instead of initializing the object with them.

Ah!.. I like that approach!.. okay will do.

Commit:
Abstract node side functionality of attachable plugins

- Expand Attacher/Detacher interfaces to break up work more
  explicitly.
- Add arguments to all functions to avoid having implementers store
  the data needed for operations.
- Expand unit tests to check that Attach, Detach, WaitForAttach,
  WaitForDetach, MountDevice, and UnmountDevice get call where
  appropriet.

swagiaal commented:
@saad-ali okay PTAL.
I have updated the functions so that they take all the arguments they _should_ need so nothing is stored in the implementer. This is a best guess, If we find out they need more or less as we add plugins to this interface I will modify them then

saad-ali commented on pkg/volume/volume.go:
```diff
@@ -129,12 +132,41 @@ type Deleter interface {
 // Attacher can attach a volume to a node.
 type Attacher interface {
```
Just realized none of the plugins actually implement the Attacher interface yet. They all just overload the Mounter code. Nice.

saad-ali commented on pkg/kubelet/kubelet.go:
```diff
@@ -140,6 +140,9 @@ const (
 	ImageGCPeriod = 5 * time.Minute
+
+	// Maximum period to wait for pod volume setup operations
+	maxWaitForVolumeOps = 5 * time.Minute
```
We'll probably want this to be longer like 20 minutes. See https://github.com/kubernetes/kubernetes/issues/15382#issuecomment-153268655

saad-ali commented on pkg/kubelet/kubelet.go:
```diff
@@ -2008,11 +2011,25 @@ func (kl *Kubelet) cleanupOrphanedVolumes(pods []*api.Pod, runningPods []*kubeco
 					glog.Errorf("Could not detach volume %q: %v", name, err)
 				}
+
+				if err = detacher.WaitForDetach(devicePath, maxWaitForVolumeOps); err != nil {
```
`WaitForDetach` should not be called here--because it blocks the sync loop. Since it is not actually implemented by any plugin yet, you could just put a comment here.

saad-ali commented:
LGTM mod nits.

Let's get this in. Like you said, we can iterate on it. There is lots of clean up work needed in the volume code, but we can continue to do that along the way and not get distracted from our goal of eliminating the racy mount/unmount behavior.

swagiaal commented:
@saad-ali PTAL

saad-ali commented:
LGTM

swagiaal commented:
@k8s-bot test this. issue #24265

swagiaal commented:
@k8s-bot test this. issue #21640

swagiaal commented:
@k8s-bot test this. issue #21640

yujuhong commented:
@swagiaal the integration test failure is not a flake, and it has nothing to do with the issues you linked.
```
+++ [0503 11:52:49] Running tests without code coverage
# k8s.io/kubernetes/test/integration
_output/local/go/src/k8s.io/kubernetes/test/integration/persistent_volumes_test.go:57: too few values in struct initializer
```

swagiaal commented:
@yujuhong yes!.. my bad.. thank you

swagiaal commented:
Integration test issue fixed. @saad-ali PTAL

swagiaal commented:
@k8s-bot test this. #IGNORE

saad-ali commented:
Taking another look at this:
```GO
Detach(deviceMountPath string, hostName string) error
```

`mountPath` doesn't make sense to pass to detach. Remember that `Detach()` could be called from the attach/detach controller (not on the node). I think we should pass in the volume spec, same as the `Attach()` method--internally a plugin should be able to use the volume spec to recreate whatever it needs to detach the volume. Thoughts?

swagiaal commented:
> I think we should pass in the volume spec, same as the Attach() method--internally a plugin should be able to use the volume spec to recreate whatever it needs to detach the volume. Thoughts?

The volume spec is not available at tear down time. Currently tear down is initiated by `cleanupOrphanedVolumes()` which looks at dereferenced mount paths.

Maybe we can add another function `GetVolSpecFromMountPath` which constructs a volume spec which basically has the disk id and change `Detach` to take a vol spec. That way if you have a volume spec available in the Attach/Detach controller you can use it directly.

But what about the case when you are populating the Attach/Detach controller cache from the cloud provider ? I guess you would need another plugin specific function which constructs a volSpec from whatever the cloud provider returns.

saad-ali commented:
> Maybe we can add another function GetVolSpecFromMountPath which constructs a volume spec which basically has the disk id and change Detach to take a vol spec. That way if you have a volume spec available in the Attach/Detach controller you can use it directly.

We could do that for the existing cleanup OrphanedVolumes code. But I doubt the entire volume spec can be recreated based on the mount path (which contains essentially the plugin and disk names).

With the attach/detach controller, I plan to store the spec in memory so I can pass it in on the detach side identical to the attach side.

> But what about the case when you are populating the Attach/Detach controller cache from the cloud provider ? I guess you would need another plugin specific function which constructs a volSpec from whatever the cloud provider returns.

Good point. The new method that Attacher plugins will implement to return all attached volumes will have to return the volume spec/node name pairs--the plugins should have enough information to construct a volume spec, and then the attacher/detacher controller can store it in-memory as normal and operate on that.

swagiaal commented:
> We could do that for the existing cleanup OrphanedVolumes code. But I doubt the entire volume spec can be recreated based on the mount path (which contains essentially the plugin and disk names).

The volume spec only needs to contain minimal information needed to perform the detach. In this case it is just the volumeID. The reason I suggested using a volume spec is because we'll need a non-plugin specific structure which the kubelet can use. On top of that we can add interface to construct the volume spec from different sources.

saad-ali commented:
> The volume spec only needs to contain minimal information needed to perform the detach.

True. I'm ok with this change. Do you want to make it or should I?

swagiaal commented:
if you could make the change that would be great.. I want to focus on the Mount/Unmount manager this week

saad-ali commented:
Will do

saad-ali commented:
Sent out PR https://github.com/kubernetes/kubernetes/pull/25523. Decided that volumeSpec was overkill and limited it to disk name which both kubelet and attach/detach controller will be able to figure out.
